rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/security/ResourceAuthenticationFilter.java

```diff
@@ -0,0 +1,107 @@
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.eclipse.sw360.rest.resourceserver.security;
+
+import org.apache.log4j.Logger;
+import org.springframework.context.annotation.Profile;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.context.SecurityContext;
+import org.springframework.security.core.context.SecurityContextHolder;
+
+import javax.servlet.*;
+import javax.servlet.http.HttpServletRequest;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@Profile("!SECURITY_MOCK")
+public class ResourceAuthenticationFilter implements Filter {
+
+    private static final Logger log = Logger.getLogger(ResourceAuthenticationFilter.class);
+    private static final String AUTHENTICATION_TOKEN_PARAMETER = "authorization";
+
+    @Override
+    public void init(FilterConfig filterConfig) {
+    }
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain filterChain) throws IOException, ServletException {
+        SecurityContext context = SecurityContextHolder.getContext();
+
+        if (context.getAuthentication() != null && context.getAuthentication().isAuthenticated()) {
+            log.debug("Already authenticated");
+        } else {
+            Map<String, String> headers = Collections.list(((HttpServletRequest) request).getHeaderNames()).stream()
+                    .collect(Collectors.toMap(h -> h, ((HttpServletRequest) request)::getHeader));
+            if (!headers.isEmpty() && headers.containsKey(AUTHENTICATION_TOKEN_PARAMETER)) {
+                String token = headers.get(AUTHENTICATION_TOKEN_PARAMETER);
+                token = token.replaceAll("(?i)token", "").trim();
+                if (token != null) {
+                    Authentication auth = new TokenAuthentication(token);
+                    SecurityContextHolder.getContext().setAuthentication(auth);
+                }
+            }
+        }
+
+        filterChain.doFilter(request, response);
+    }
+
+    @Override
+    public void destroy() {
+    }
+
+    class TokenAuthentication implements Authentication {
+        private String token;
+
+        private TokenAuthentication(String token) {
+            this.token = token;
+        }
+
+        @Override
+        public Collection<? extends GrantedAuthority> getAuthorities() {
+            return new ArrayList<>(0);
+        }
+
+        @Override
+        public Object getCredentials() {
+            return token;
+        }
+
+        @Override
+        public Object getDetails() {
+            return null;
+        }
+
+        @Override
+        public Object getPrincipal() {
+            return null;
```

Comment:
ping @maierthomas : I think this comment here is the last without code change or response

Reply:
In case of  API Tokens (Authorization: Token) `ApiTokenAuthentication` is used.
The only information is stored in the `token` field.
All other fields like principals, details, authorities, authenticated are determined and filled in the authentication provider.
For every `ApiTokenAuthentication` a look up in the user data db (view - getApiToken) is required.

Reply:
So there is some magic thing which exchanges the `ApiTokenAuthentication` in the context, which was set via `.setAuthentication(auth)`, with some other `? implements Authentication`, which has a valid `.getPrincipal` implementation? Can you point me to that magic functionallity, I am not able to find it.

Reply:
It creates a new custom authentication named `ApiTokenAuthentication` in the the filter chain and the provider class `ResourceAuthenticationProvider` verifies the token, validity, scopes..
`ApiTokenAuthentication` implements Spring `Authentication` and stores only the token value as `principle`. To provide `ApiTokenAuthentication` to the authentication provider it needs to be set to the security context (setAuthentication).

Json Web Token are already set to the security context and are authenticated. The `ResourceAuthenticationFilter` just forwards the authentication to the next filter `filterChain.doFilter(request, response);`